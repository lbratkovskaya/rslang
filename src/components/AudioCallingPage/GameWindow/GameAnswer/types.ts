import { IWord } from '../../../../store/types';

export default interface IAnswerProps {
  index: number;
  increment: () => void;
  track: HTMLAudioElement;
  words: IWord[];
}
