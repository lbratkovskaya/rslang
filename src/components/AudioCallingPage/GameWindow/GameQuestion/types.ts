export default interface IQuestionsProps {
  play: () => void;
  track: HTMLAudioElement;
}
