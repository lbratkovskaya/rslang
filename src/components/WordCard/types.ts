import { IWord } from '../../store/types';

export interface IWordCardProps {
  word: IWord;
  index: number;
  activeGroup: any;
  isLoading: boolean;
  showDeleted: boolean;
  removeOnDifficultyChange: boolean;
}

export interface IWordCardButton {
  label: string;
  altLabel: string;
  title: string;
  altTitle: string;
  clickable: boolean;
  onClick?: () => void;
  param?: boolean;
}
