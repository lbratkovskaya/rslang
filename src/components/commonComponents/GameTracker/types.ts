export default interface IGameTracker {
  index: number;
  end: number;
}
