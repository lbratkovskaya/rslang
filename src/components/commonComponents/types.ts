import { FullScreenHandle } from 'react-full-screen';
import { IGroup } from '../../constants';
import { IAudioCallingWords } from '../../store/reducers/audioCallingReducer/types';
import { ISavannahWord } from '../../store/types';

export interface IGameSelectProps {
  changeSelectFc: (item: string | number) => void;
  selectData: Array<IGroup> | number;
  selectName: string;
  disabled: boolean;
  value: number;
}

export interface ILangSwitchProps {
  handleSwitch: (checked: boolean) => void;
  isLang: boolean;
}

export interface IFullScreenBtnProps {
  changeScreen: () => void;
}

export interface IGameExitBtnProps {
  clickBtn: () => void;
}

export interface IFullScreenWrapperProps {
  component: {};
  handle: FullScreenHandle;
}

export interface IRenderTimeProps {
  remainingTime: number;
}

export interface ITimerProps {
  gameTime: number;
  size: number;
  handleOnComplite: () => void;
}

export interface IGameModeProps {
  changeModeFc: (mode: string) => void;
}

export interface ITableEndProps {
  words: Array<ISavannahWord | IAudioCallingWords>;
}
