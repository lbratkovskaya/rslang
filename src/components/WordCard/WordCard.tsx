import React, { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Transition, TransitionStatus } from 'react-transition-group';
import Parser from 'html-react-parser';
import { Typography, Card, Chip, useTheme, Menu, MenuItem, Tooltip } from '@material-ui/core';
import { Done, VolumeUpRounded, StopRounded, ChevronLeft } from '@material-ui/icons';
import {
  setUserWordDeleted,
  setUserWordEasy,
  setUserWordHard,
  deleteUserWord,
} from '../../store/actions/dictionaryActions';
import { deleteWordFromGamesStore } from '../../store/actions/gamesActions';
import backendUrl, {
  APPEAR_DURATION,
  APPEAR_STYLE,
  WORDBOOK_GROUPS,
  WORDCARD_APPEAR_GAP,
} from '../../constants';
import { IAppState, IUserWord } from '../../store/types';
import { IWordCardButton, IWordCardProps } from './types';
import useStyles, { defaultImageSize, transitionStyles } from './styles';

const WordCard: React.FC<IWordCardProps> = ({
  word,
  index,
  activeGroup,
  isLoading,
  showDeleted,
  removeOnDifficultyChange,
}: IWordCardProps) => {
  const classes = useStyles();
  const [isMounted, setIsMounted] = useState(false);
  const { data: userData } = useSelector((state: IAppState) => state.user);
  const isUserLoggedIn = useSelector((state: IAppState) => state.user.isLoggedIn);
  const userDifficultWords = useSelector((state: IAppState) =>
    state.userDictionary.difficultWords.map((el) => el.word)
  );
  const isInDifficultWords = userDifficultWords.includes(word.word);
  const userDeletedWords = useSelector((state: IAppState) =>
    state.userDictionary.deletedWords.map((el) => el.word)
  );
  const isInDeletedWords = userDeletedWords.includes(word.word);

  const userWords =
    useSelector((state: IAppState) =>
      state.userDictionary.learningWords?.reduce((acc, el) => {
        Object.assign(acc, { [el.word]: el });
        return acc;
      }, {} as { [key: string]: IUserWord })
    ) || {};
  const userWord = userWords[word.word] || word;

  const { showTranslate, showButtons } = useSelector((state: IAppState) => state.wordBook);
  const [isImageReady, setImageIsReady] = useState(false);
  const audio = useMemo(() => new Audio(), []);
  const [isAudioPlaying, setIsAudioPlaying] = useState(false);
  const [playingAudioIndex, setPlayingAudioIndex] = useState(-1);
  const [isDeleted, setIsDeleted] = useState(false);
  const [isDifficult, setIsDifficult] = useState(false);
  const [restoreAnchor, setRestoreAnchor] = useState<null | HTMLElement>(null);
  const { color } = WORDBOOK_GROUPS[activeGroup] || 'rgb(255, 0, 0)';
  const highlightStyle = { color };
  const theme = useTheme();
  const colorOfDifficult = theme.palette.secondary.main;
  const dispatch = useDispatch();

  const textStyle = {
    word: playingAudioIndex === 0 ? highlightStyle : {},
    meaning: playingAudioIndex === 1 ? highlightStyle : {},
    example: playingAudioIndex === 2 ? highlightStyle : {},
  };

  const difficultStyle = {
    filter: isDifficult ? `drop-shadow(0 0 3px ${colorOfDifficult})` : '',
  };

  const preloadImage = (): void => {
    const img = new Image();
    img.onload = (): void => setImageIsReady(true);
    img.src = `${backendUrl}/${word.image}`;
  };

  const playAudio = (src: string, audioIndex: number): void => {
    setIsAudioPlaying(true);
    setPlayingAudioIndex(audioIndex);
    audio.src = `${backendUrl}/${src}`;
    audio.play();
  };

  const renderButton = ({
    label,
    altLabel,
    title,
    altTitle,
    clickable,
    onClick,
    param,
  }: IWordCardButton): JSX.Element => {
    const userTitle = param ? altTitle : title;
    const chipTitle = isUserLoggedIn ? userTitle : 'Авторизуйтесь, чтобы воспользоваться функцией';
    return (
      <>
        <Tooltip title={chipTitle} placement="bottom">
          <Chip
            className={`${classes.button} ${!isUserLoggedIn && classes.buttonInactive}`}
            variant={param ? 'default' : 'outlined'}
            color={param ? 'secondary' : 'default'}
            size="small"
            deleteIcon={param ? <Done /> : <></>}
            clickable={clickable}
            label={param ? altLabel : label}
            onClick={clickable ? onClick : undefined}
            onDelete={clickable ? onClick : undefined} // necessary for deleteIcon to be rendered
          />
        </Tooltip>
      </>
    );
  };

  const handleAudioClick = (): void => {
    playAudio(word.audio, 0);
    audio.onended = () => {
      playAudio(word.audioMeaning, 1);
      audio.onended = () => {
        playAudio(word.audioExample, 2);
        audio.onended = () => {
          setIsAudioPlaying(false);
          setPlayingAudioIndex(-1);
        };
      };
    };
  };

  const handleStopClick = (): void => {
    audio.pause();
    setIsAudioPlaying(false);
    setPlayingAudioIndex(-1);
  };

  const handleAddToDifficult = (): void => {
    if (isDifficult) {
      dispatch(setUserWordEasy(userWord, userData));
      setIsDifficult(false);
    } else {
      dispatch(setUserWordHard(userWord, userData));
      setIsDifficult(true);
    }
    if (removeOnDifficultyChange) {
      setIsMounted(false);
    }
  };

  const handleDelete = (): void => {
    if (isDeleted) {
      return;
    }
    dispatch(setUserWordDeleted(userWord, userData, !isDeleted));
    dispatch(deleteWordFromGamesStore(userWord));
    setIsMounted(false);
    setTimeout(() => {
      setIsDeleted(true);
    }, APPEAR_DURATION);
  };

  const handleRestoreClick = (event: React.MouseEvent<HTMLElement>): void => {
    setRestoreAnchor(event.currentTarget);
  };

  const handleRestoreClose = (): void => {
    setRestoreAnchor(null);
  };

  const handleKeepInDictionary = (): void => {
    handleRestoreClose();
    dispatch(setUserWordDeleted(userWord, userData, !isDeleted));
    setIsMounted(false);
    setTimeout(() => {
      setIsDeleted(true);
    }, APPEAR_DURATION);
  };

  const handleRemoveFromDictionary = (): void => {
    handleRestoreClose();
    dispatch(deleteUserWord(userWord, userData));
    setIsMounted(false);
    setTimeout(() => {
      setIsDeleted(true);
    }, APPEAR_DURATION);
  };

  const renderMainParagraph = (title: string, content: string, className: {}): JSX.Element => (
    <Typography variant="body2">
      {title}: <span style={className}>{Parser(content)}</span>
    </Typography>
  );

  const renderParagraph = (title: string, content: string): JSX.Element => (
    <Typography variant="body2" color="textSecondary" className={classes.secondary}>
      {title}: {Parser(content)}
    </Typography>
  );

  const wordTranslateComponent = showTranslate && (
    <span className={classes.wordTranslate}>{` — ${word.wordTranslate}`}</span>
  );

  useEffect(() => {
    const delay = WORDCARD_APPEAR_GAP * index;
    const cardAppearTimeout = setTimeout(() => setIsMounted(true), delay);
    preloadImage();

    if (userDifficultWords.includes(word.word)) setIsDifficult(true);
    if (userDeletedWords.includes(word.word)) setIsDeleted(true);

    return () => {
      clearTimeout(cardAppearTimeout);
      setIsMounted(false);
      handleStopClick();
    };
  }, []);

  useEffect(() => {
    setIsDifficult(isInDifficultWords);
  }, [isInDifficultWords]);

  useEffect(() => {
    setIsDeleted(isInDeletedWords);
  }, [isInDeletedWords]);

  return (
    <Transition in={isMounted && !isLoading} timeout={APPEAR_DURATION} unmountOnExit>
      {(state: TransitionStatus) => (
        <>
          {(!isDeleted || showDeleted) && (
            <Card
              className={classes.card}
              style={{ ...APPEAR_STYLE, ...difficultStyle, ...transitionStyles[state] }}>
              <img
                src={`${backendUrl}/${word.image}`}
                alt={word.word}
                width={defaultImageSize.width}
                height={defaultImageSize.height}
                className={classes.image}
                style={isImageReady ? { opacity: 1 } : {}}
              />
              <Typography className={classes.word}>
                <span style={textStyle.word}>{word.word}</span>
                {wordTranslateComponent}
              </Typography>
              <Typography className={classes.transcription}>
                {` ${word.transcription} `}
                <VolumeUpRounded
                  onClick={handleAudioClick}
                  color={isAudioPlaying ? 'disabled' : 'action'}
                  className={classes.icon}
                />
                {isAudioPlaying && (
                  <StopRounded
                    onClick={handleStopClick}
                    color="secondary"
                    className={classes.icon}
                  />
                )}
              </Typography>
              {renderMainParagraph('Meaning', word.textMeaning, textStyle.meaning)}
              {showTranslate && renderParagraph('Значение', word.textMeaningTranslate)}
              {renderMainParagraph('Example', word.textExample, textStyle.example)}
              {showTranslate && renderParagraph('Пример', word.textExampleTranslate)}
              {showButtons &&
                !isDeleted &&
                renderButton({
                  label: 'Добавить в сложные',
                  altLabel: 'Добавлено в сложные',
                  title: 'Добавить слово в список сложных слов',
                  altTitle: 'Удалить слово из списка сложных слов',
                  clickable: isUserLoggedIn,
                  onClick: handleAddToDifficult,
                  param: isDifficult,
                })}
              {showButtons &&
                renderButton({
                  label: 'В удалённые',
                  altLabel: 'Удалено',
                  title: 'Переместить слово в удалённые',
                  altTitle: '',
                  clickable: isUserLoggedIn && !isDeleted,
                  onClick: handleDelete,
                  param: isDeleted,
                })}
              {showButtons && isUserLoggedIn && isDeleted && (
                <>
                  <Chip
                    aria-controls="simple-menu"
                    className={classes.button}
                    variant="outlined"
                    color="default"
                    size="small"
                    clickable
                    label="Восстановить"
                    onClick={handleRestoreClick}
                  />
                  <Menu
                    id="simple-menu"
                    anchorEl={restoreAnchor}
                    anchorOrigin={{
                      vertical: 'bottom',
                      horizontal: 'left',
                    }}
                    getContentAnchorEl={null}
                    keepMounted
                    open={Boolean(restoreAnchor)}
                    onClose={handleRestoreClose}>
                    <MenuItem onClick={handleRemoveFromDictionary}>Удалить из словаря</MenuItem>
                    <MenuItem onClick={handleKeepInDictionary}>Оставить в словаре</MenuItem>
                  </Menu>
                </>
              )}
              {userWords[word.word]?.userWord && (
                <div className={classes.heatsPanel}>
                  <ChevronLeft className={classes.chevron} />
                  <Typography className={classes.successHeats}>
                    Попаданий: {userWords[word.word]?.userWord?.optional.successHeats || 0}
                  </Typography>
                  <Typography className={classes.errorHeats}>
                    Промахов: {userWords[word.word]?.userWord?.optional.errorHeats || 0}
                  </Typography>
                </div>
              )}
            </Card>
          )}
        </>
      )}
    </Transition>
  );
};

export default WordCard;
