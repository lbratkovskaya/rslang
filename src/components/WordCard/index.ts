import WordCard from './WordCard';

export default WordCard;
