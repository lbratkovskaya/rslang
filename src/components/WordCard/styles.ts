import { createStyles, makeStyles, StyleRules, Theme } from '@material-ui/core';
import { APPEAR_DURATION } from '../../constants';

const COLORS = {
  secondaryText: 'rgba(0, 0, 0, 0.35)',
};

export const wordCardStyles = (theme: Theme): StyleRules => ({
  card: {
    margin: theme.spacing(1),
    padding: theme.spacing(1),
    minWidth: 450,
    textAlign: 'left',
    position: 'relative',
    height: '100%',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
      maxWidth: 600,
    },
  },
  word: {
    textTransform: 'capitalize',
    fontWeight: 500,
  },
  transcription: {
    marginBottom: theme.spacing(1),
    display: 'flex',
    color: COLORS.secondaryText,
    lineHeight: 2,
    fontSize: 14,
    fontWeight: 400,
    letterSpacing: 1,
  },
  icon: {
    cursor: 'pointer',
    transition: 'color 0.2s',
    '&:hover': {
      color: 'black',
    },
  },
  button: {
    marginRight: theme.spacing(1),
    transition: 'background 0.5s, color 0.5s',
  },
  buttonInactive: {
    color: COLORS.secondaryText,
  },
  wordTranslate: {
    textTransform: 'capitalize',
    fontWeight: 400,
  },
  secondary: {
    marginBottom: theme.spacing(1),
  },
  image: {
    float: 'right',
    marginLeft: theme.spacing(1),
    borderRadius: theme.spacing(1) / 2,
    opacity: 0,
    transition: `opacity ${APPEAR_DURATION}ms`,
  },
  heatsPanel: {
    position: 'absolute',
    display: 'flex',
    minWidth: '130px',
    minHeight: '100%',
    padding: '8px',
    paddingLeft: '30px',
    flexDirection: 'column',
    justifyContent: 'center',
    gap: '30px',
    right: 0,
    top: 0,
    background: 'rgba(255, 255, 255, .8)',
    transition: 'transform .3s',
    transform: 'translateX(calc(100% - 30px))',
    '&:hover': {
      transform: 'translateX(0)',
    },
  },
  chevron: {
    position: 'absolute',
    left: 0,
    transform: 'scale(1.3)',
    color: 'rgba(0, 0, 0, 0.47)',
  },
  successHeats: {
    fontSize: '1.2rem',
    fontWeight: 500,
    color: '#1b861b',
  },
  errorHeats: {
    fontSize: '1.2rem',
    fontWeight: 500,
    color: '#c01818',
  },
});

const useStyles = makeStyles((theme: Theme) => createStyles(wordCardStyles(theme)));

export const transitionStyles: { [key: string]: {} } = {
  entering: { opacity: 0 },
  entered: { opacity: 1 },
  exiting: { opacity: 0, transform: 'scale(0)' },
  exited: { opacity: 0 },
};

export const defaultImageSize = {
  width: 240,
  height: 160,
};

export default useStyles;
