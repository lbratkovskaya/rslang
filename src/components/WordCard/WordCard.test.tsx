import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallow, ShallowWrapper } from 'enzyme';
import WordCard from './WordCard';
import { IWord } from '../../store/types';
import store from '../../store';

const mockStore = configureStore([]);

describe('<WordCard /> testing:', () => {
  let wrapper: ShallowWrapper<typeof WordCard>;

  const testStore = mockStore(store);

  beforeEach(() => {
    wrapper = shallow<typeof WordCard>(
      <Provider store={testStore}>
        <WordCard
          word={{} as IWord}
          index={0}
          activeGroup={0}
          isLoading={false}
          showDeleted
          removeOnDifficultyChange={false}
        />
      </Provider>
    );
  });

  it('Must have a correct snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('Should render a single element named WordCard', () => {
    expect(wrapper.find('WordCard')).toHaveLength(1);
  });
});
