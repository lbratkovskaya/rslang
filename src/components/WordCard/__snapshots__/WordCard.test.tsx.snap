// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WordCard /> testing: Must have a correct snapshot 1`] = `
"<ContextProvider value={{...}}>
  <WordCard word={{...}} index={0} activeGroup={0} isLoading={false} showDeleted={true} removeOnDifficultyChange={false} />
</ContextProvider>"
`;
